src/sdi-progress-dock.c

```diff
@@ -0,0 +1,97 @@
+#include <snapd-glib/snapd-glib.h>
+
+/**
+ * This class manages the progress bars in the dock for each snap
+ * being updated.
+ */
+
+struct _SdiProgressDock {
+  GObject parent_instance;
+
+  GApplication *application;
+  UnityComCanonicalUnityLauncherEntry *unity_manager;
+};
+
+G_DEFINE_TYPE(SdiProgressDock, sdi_progress_dock, G_TYPE_OBJECT)
+
+/**
+ * This callback should be connected to the `refresh-progress` signal from a
+ * #sdi_refresh_monitor object. It will receive the total number of tasks and
+ * how many have been done, and if the task has been completed, and with that
+ * will update the progress bars in the dock.
+ */
+void sdi_progress_dock_update_progress(SdiProgressDock *self, gchar *snap_name,
+                                       GStrv desktop_files,
+                                       gchar *task_description,
+                                       guint done_tasks, guint total_tasks,
+                                       gboolean task_done) {
+  if (desktop_files != NULL && total_tasks != 0) {
+    for (gchar **desktop_file = desktop_files; *desktop_file != NULL;
+         desktop_file++) {
+      // Update dock progress bar
+      g_autoptr(GVariantBuilder) builder =
+          g_variant_builder_new(G_VARIANT_TYPE("a{sv}"));
+      g_variant_builder_add(
+          builder, "{sv}", "progress",
+          g_variant_new_double(done_tasks / ((gfloat)total_tasks)));
+      g_variant_builder_add(builder, "{sv}", "progress-visible",
+                            g_variant_new_boolean(!task_done));
+      g_variant_builder_add(builder, "{sv}", "updating",
+                            g_variant_new_boolean(!task_done));
+
+      g_autoptr(GVariant) values =
+          g_variant_ref_sink(g_variant_builder_end(builder));
+      unity_com_canonical_unity_launcher_entry_emit_update(
+          self->unity_manager, *desktop_file, values);
+    }
+  }
+}
+
+static void sdi_progress_dock_dispose(GObject *object) {
+  SdiProgressDock *self = SDI_PROGRESS_DOCK(object);
+
+  g_clear_object(&self->unity_manager);
+  g_clear_object(&self->application);
+
+  G_OBJECT_CLASS(sdi_progress_dock_parent_class)->dispose(object);
+}
+
+static void sdi_progress_dock_class_init(SdiProgressDockClass *klass) {
+  GObjectClass *gobject_class = G_OBJECT_CLASS(klass);
+
+  gobject_class->dispose = sdi_progress_dock_dispose;
+}
+
+static void sdi_progress_dock_init(SdiProgressDock *self) {}
+
+SdiProgressDock *sdi_progress_dock_new(GApplication *application) {
+  SdiProgressDock *self = g_object_new(SDI_TYPE_PROGRESS_DOCK, NULL);
+  self->application = g_object_ref(application);
+  g_autofree gchar *unity_object =
+      g_strdup_printf("/com/canonical/unity/launcherentry/%d", getpid());
+  self->unity_manager = unity_com_canonical_unity_launcher_entry_skeleton_new();
+  g_dbus_interface_skeleton_export(
+      G_DBUS_INTERFACE_SKELETON(self->unity_manager),
+      g_application_get_dbus_connection(application), unity_object, NULL);
```

Comment:
I feel we should handle this error, at least by logging it.

You can also return accordingly, so that caller is aware of the failure.

If you want the clean solution instead, you should implement this as a `GInitable`.

Reply:
I added a login warning. Check it when you have some spare time, please.